### PR TITLE
Branch sample post and prior

### PR DESCRIPTION
**Pull request by dngoldberg**
**What feature does this add?**

This P/R predominantly adds a the run script  `fenics_ice/runs/run_sample.py`, which enables sampling **zero-mean** realisations from the _prior_ and _posterior_ distributions of `alpha` and `beta`. 

What this means is that e.g. `alpha` is represented by a random vector with gaussian distribution given by mean alpha_0 and covariance Gamma_0, i.e. the prior distribution is N (alpha_0, Gamma_0). This pull request enables a _sample_ of from the distribution, but with **zero mean**, i.e. N( 0, Gamma_0 ). Hence the effects of the covariance matrix can be more clearly seen. (By default, i believe our alpha_0 = 0, but this should be checked. **Note alpha_0 is not the initial guess for alpha!**)

The prior sample is generated by generating an array of independent standard normals of the size of alpha (or beta), and left-mutiplying by the _square root_ of the prior covariance. Thus, if the prior covariance is L M^{-1} L, the square root is L M^{-1/2}.

There is an option to additionally sample the _posterior_ distribution (see Koziol et al, 2021, 5.3.1 for a description of the mathematics involved). Again, this samples form a distribution with the posterior covariance, but zero mean, allowing for a better visualisation of posterior uncertainty. Note that if the posterior is sampled, this assumes a completed eigendecomposition, the location of which is indicated by the toml. The run will fail if none is found.

The run script uses `write_variable` to save the relevant diagnostics files to the appropriate directory.

A new section `sample` has been added to the .toml file, with parameters:

```python
sample_size: int = 1                            # The number of samples to take -- if larger than 1, the mean and standard deviation over the samples is found. not sure this makes sense to do but it was included.
sample_alpha: bool = False                # choose whether to sample alpha or not.
sample_beta: bool = False                 # choose whether to sample beta or not.
sample_posterior: bool = False          # choose whether to sample the posterior or not. 
num_eigenvals: int = 0                       # The maximum number of eigenvalues retained in the posterior sample. If zero or larger than the full number of eigenvalues, the full number is retained (subject to eigenvalue_thresh). 
phase_name: str = 'sample'               # parameter to determine save location for diagnostics
phase_suffix: str = ''                           # parameter to determine suffix for diagnostics
```

In addition, `prior.py` is modified to find the prior square root and inverse sq root action, `sqrt_matrix_action.py` was added to `fenics_ice` to enable the square root of the mass matrix, and `config.py` was modified as above.

_**NOTE: in sqrt_matrix_action.py the maximum iter count was set to 1e5. This is much larger than that needed in the idealised problem in Koziol et al, 2021, GMD.**_

### CODE REVIEW

**Comment by jrmaddison** [comment]
> _**NOTE: in sqrt_matrix_action.py the maximum iter count was set to 1e5. This is much larger than that needed in the idealised problem in Koziol et al, 2021, GMD.**_

Perhaps try `LumpedPCSqrtMassAction`.

**Comment by dngoldberg** [comment]
thanks @jrmaddison -- just so im certain, the lumped aspect is limited to the preconditioner?

**Comment by jrmaddison** [comment]
> thanks @jrmaddison -- just so im certain, the lumped aspect is limited to the preconditioner?

Yes, that's correct. This uses the lumped mass matrix to precondition the square root mass action, but still yields samples from the Gaussian random variable defined using the full mass matrix. You should be able to choose a much larger value of `beta`.

**Comment by jrmaddison** [comment]
Likely due to the use of a mixed function space. `M_L = assemble(sum(split(test), ufl.zero()) * dx)` should work more generally.

**Comment by dngoldberg** [comment]
converted to draft only until i can check the outputs in paraview after converting to lumped preconditioned mass matrix inverse

**Comment by dngoldberg** [comment]
@jrmaddison i have tested with the preconditioned sqrt mass matrix solver and the samples that result are identically zero (they are not meant to be). With the only difference being the use of LumpedPCSqrtMassAction.action() instead of A_root_action(), possible explanations are that Im not instantiating LumpedPCSqrtMassAction correctly, or there is a bug within LumpedPCSqrtMassAction. In the latest commit i have left my instantiation of LumpedPCSqrtMassAction in prior.py but have used A_root_action(). If you see anything obvious let me know, otherwise can i suggest this be left aside so the rest of the PR can be reviewed.

File: fenics_ice/prior.py [hunk 1]
```diff
@@ -80,6 +82,13 @@ def __init__(self, slvr, space):
         self.construct_mass_operator()
         self.construct_prior_operator()
 
+        self.tmp1, self.tmp2, self.tmp3 = Vector(), Vector(), Vector()
+        self.A.init_vector(self.tmp1, 0)
+        self.A.init_vector(self.tmp2, 1)
+
+        # preconditioned solver object to find square root of mass matrix (not used)
+        self.lumpedPCMassSolver = LumpedPCSqrtMassAction(space=self.space, tol=1.0e-16)
```
Comment (jrmaddison) [inline fenics_ice/prior.py:90]: Unused.

File: fenics_ice/prior.py [hunk 2]
```diff
@@ -291,6 +300,23 @@ def inv_action(self, x, y):
         y.set_local(self.tmp1.get_local())
         y.apply("insert")
 
+    def sqrt_action(self,x,y):  # sqrt of inv cov: Gamma -1 Gamma 1/2
+                                #                  L M-1 L L-1 M1/2
+                                #                  L M-1 M1/2
+        M_norm = self.M.norm("linf")
+        self.tmp1, terms = A_root_action(self.M, x, tol=1.0e-16, beta=M_norm, max_terms=100000)
```
Comment (jrmaddison) [inline fenics_ice/prior.py:307]: The tolerance is very small

File: fenics_ice/prior.py [hunk 3]
```diff
@@ -291,6 +300,23 @@ def inv_action(self, x, y):
...
+        self.M_solver.solve(self.tmp2, self.tmp1) 
+        self.A.mult(self.tmp2,self.tmp3)
+        y.set_local(self.tmp3.get_local())
+        y.apply("insert")
```
Comment (jrmaddison) [inline fenics_ice/prior.py:311]: Can use `y` in place of `self.tmp3` here

File: fenics_ice/prior.py [hunk 4]
```diff
@@ -291,6 +300,23 @@ def inv_action(self, x, y):
...
+
+    def sqrt_inv_action(self,x,y):  # sqrt of inv cov: Gamma 1/2
+                                    #                  L-1 M1/2
+        M_norm = self.M.norm("linf")        
+        self.tmp1, terms = A_root_action(self.M, x, tol=1.0e-16, beta=M_norm, max_terms=100000)
```
Comment (jrmaddison) [inline fenics_ice/prior.py:316]: The tolerance is very small

File: fenics_ice/sqrt_matrix_action.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+        A_root_action.
+        Arguments:
+            space      The function space.
+            tol, beta  As for A_root_action.
+            M          (Optional) Mass matrix.
+        """
+
+        test = TestFunction(space)
+        M_L = assemble(sum(split(test), ufl.zero()) * dx) 
+        #M_L = assemble(test * dx)
```
Comment (jrmaddison) [inline fenics_ice/sqrt_matrix_action.py:93]: Commented-out code

**Comment by dngoldberg** [comment]
thanks for the support @jrmaddison - i put in the additional apply('insert') and also fixed the bug i had with LumpedPCSqrtMassSolver().

File: fenics_ice/prior.py (hunk 1 above)
Comment (dngoldberg) [reply fenics_ice/prior.py]: addressed now

File: fenics_ice/prior.py (hunk 2 above)
Comment (dngoldberg) [reply fenics_ice/prior.py]: with the lumped PC it is much faster, so likely OK

File: runs/run_sample.py [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+    if (sample_posterior):
+     am.vector().set_local(np.zeros(shp))
+     am.vector().apply("insert")
+     if (ssize>1):
+      astd.vector().set_local(np.zeros(shp))
+      astd.vector().apply("insert")
+
+    for i in range(params.sample.sample_size):
+
+      np.random.seed()
```
[the remark this answers is not in the document]
Comment (dngoldberg) [reply runs/run_sample.py:175]: would you suggest calling it before the loop?

[the remark this answers is not in the document]
Comment (jrmaddison) [reply runs/run_sample.py:175]: Yes, just once before the loop should be fine

**Comment by bearecinos** [comment]
@dngoldberg if all the test pass I think we should merge it ? have you ran all of them?

**Review by bearecinos — APPROVED** [review]

**Comment by dngoldberg** [comment]
hi @bearecinos i have not run every possible combination but i believe i have tested the functionality sufficiently -- i've run without posterior, as well as with posterior and  limiting the number of eigenvalues, limiting to alpha, and with a sample size of larger than one. i will check over again to be sure, and merge.

**Comment by bearecinos** [comment]
Hi Dan ... no I meant the pytest?

**Comment by dngoldberg** [comment]
ah -- no i have not done this yet. will do!

**Pull request merged**